### PR TITLE
Blocksparse block-data caching

### PR DESCRIPTION
**Pull request by mganahl**
This PR adds a simple mechanism for caching support for block-sparse tensor contractions.
The user can turn on block-data caching by using `tn.block_sparse.enable_caching()` and turn it off
using `tn.block_sparse.disable_caching()`. 

Caching can substantially reduce run times for block-sparse tensor contractions, at the cost of a substantially larger memory footprint, and possibly catastrophic memory clutter. 

By default caching is disabled. 
This PR depends on changes in PR #733

### CODE REVIEW

File: tensornetwork/backends/symmetric/symmetric_backend.py [hunk 1]
```diff
@@ -205,11 +207,22 @@ def eigsh_lanczos(self,
         to check convergence.
       reorthogonalize: If `True`, Krylov vectors are kept orthogonal by
         explicit orthogonalization (more costly than `reorthogonalize=False`)
+      enable_caching: If `True`, block-data during calls to `matvec` is cached
+        for later reuse. Note: usually it is save to enable_caching, unless 
+        `matvec` uses matrix decompositions liek SVD, QR, eigh, eig or similar.
```
Comment (Thenerdstation) [inline tensornetwork/backends/symmetric/symmetric_backend.py:212]: liek -> like

File: tensornetwork/block_sparse/utils.py [hunk 2]
```diff
@@ -449,11 +450,23 @@ def _find_diagonal_sparse_blocks(
   return block_maps, obj, block_dims
 
 
+def _compute_hash(charges, flows, tr_partition, order):
+  """
+  hash the input arguments of _find_transposed_diagonal_sparse_blocks 
+  and return the value
+  """
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/utils.py:457]: Fully document this.

File: tensornetwork/block_sparse/utils.py [hunk 3]
```diff
@@ -478,6 +491,12 @@ def _find_transposed_diagonal_sparse_blocks(
     block_dims (np.ndarray): 2-by-m array of matrix dimensions of each block.
   """
   flows = np.asarray(flows)
+  cacher = get_cacher()
+  if cacher.do_caching:
+    hash_val = _compute_hash(charges, flows, tr_partition, order)
+    if hash_val in cacher.cache:
+      return cacher.cache[hash_val]
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/utils.py:498]: How will you deal with hash collisions?

Comment (mganahl) [reply tensornetwork/block_sparse/utils.py:498]: TBH I'm not sure

Comment (mganahl) [reply tensornetwork/block_sparse/utils.py:498]: could store the original value to be hashed, and compare any new object to be hashed to the stored object

Comment (Thenerdstation) [reply tensornetwork/block_sparse/utils.py:498]: Well, having two of these collide would be a hugely annoying bug for the end user, so this'll need to be solved.

Comment (mganahl) [reply tensornetwork/block_sparse/utils.py:498]: actually, I can just do without hashing

Comment (mganahl) [reply tensornetwork/block_sparse/utils.py:498]: fixed

File: tensornetwork/block_sparse/utils.py [hunk 4]
```diff
@@ -450,10 +450,22 @@ def _find_diagonal_sparse_blocks(
...
+    charges: List of `BaseCharge`, one for each leg of a tensor. 
+    flows: A list of bool, one for each leg of a tensor.
+      with values `False` or `True` denoting inflowing and 
+      outflowing charge direction, respectively.
+    tr_partition: Location of the transposed tensor partition 
+    (i.e. such that the tensor is viewed as a matrix between 
+    `charges[order[:partition]]` and `charges[order[partition:]]`).
+    order: Order with which to permute the tensor axes. 
+  Returns:
+    str: the string representation of the input
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/utils.py:468]: It returns a tuple and not just a string.

Comment (mganahl) [reply tensornetwork/block_sparse/utils.py]: fixed

**Comment by mganahl** [comment]
@gevenbly, this PR implements the caching strategy I discussed with you. Right now caching is used within the sparse solvers of the block-sparse backend. I think for regular use cases it should be save to use it there, and it can give good performance boosts whenever sparse solvers are the computational bottleneck. Can you have a quick look as well and let me know what you think? thanks

File: tensornetwork/backends/symmetric/symmetric_backend.py [hunk 5]
```diff
@@ -205,14 +207,20 @@ def eigsh_lanczos(self,
         to check convergence.
       reorthogonalize: If `True`, Krylov vectors are kept orthogonal by
         explicit orthogonalization (more costly than `reorthogonalize=False`)
+      enable_caching: If `True`, block-data during calls to `matvec` is cached
+        for later reuse. Note: usually it is save to enable_caching, unless
```
Comment (Thenerdstation) [inline tensornetwork/backends/symmetric/symmetric_backend.py:211]: save -> safe

File: tensornetwork/backends/symmetric/symmetric_backend.py [hunk 6]
```diff
@@ -205,14 +207,20 @@ def eigsh_lanczos(self,
         to check convergence.
       reorthogonalize: If `True`, Krylov vectors are kept orthogonal by
         explicit orthogonalization (more costly than `reorthogonalize=False`)
+      enable_caching: If `True`, block-data during calls to `matvec` is cached
+        for later reuse. Note: usually it is save to enable_caching, unless 
+        `matvec` uses matrix decompositions like SVD, QR, eigh, eig or similar.
+        In this case, if one does a large number of krylov steps, this can lead 
+        to memory clutter and/or overflow.
```
Comment (Thenerdstation) [inline tensornetwork/backends/symmetric/symmetric_backend.py:214]: and/or OOM errors.

File: tensornetwork/block_sparse/caching_test.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  L = BlockSparseTensor.random(Linds)
+  R = BlockSparseTensor.random(Rinds)
+
+  enable_caching()
+  ncon([L, mps, mpo, R], [[3, 1, -1], [1, 2, 4], [3, 5, -2, 2], [5, 4, -3]],
+       backend='symmetric')
+  cacher = get_cacher()
+  assert len(cacher.cache) > 0
+  clear_cache()
+  assert len(cacher.cache) == 0
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/caching_test.py:81]: These checks are somewhat weak. Can you also verify the data in the cache is correct?

**Comment by mganahl** [comment]
added more extensive testing and fixed some bugs

**Pull request merged**